### PR TITLE
Show categories in frontend and sort in backend

### PR DESCRIPTION
**Pull request by stedo005**

### CODE REVIEW

File: backend/src/main/java/com/example/demo/soldItems/SoldItemService.java [hunk 1]
```diff
@@ -72,31 +76,46 @@ private SoldItem makeSoldItem(HelloCashItem item, HelloCashInvoice invoice) {
 
     }
 
-    public void getItemByDate(String dateFrom, String dateTo) {
+    public List<SoldItem> getItemByQueryData(String categoryId, String dateFrom, String dateTo) {
+
+        List<String> itemsInCategory;
+
+        if (categoryRepository.findById(categoryId).isPresent()) {
```
Comment (gossie) [inline backend/src/main/java/com/example/demo/soldItems/SoldItemService.java:83]: You are doing that call twice: `categoryRepository.findById(categoryId)` that means there are two requests that are sent to the database.
You can reduce it to one doing this:
```java
List<String> itemsInCategory = categoryRepository.findById(categoryId)
        .map(category -> category.getItemsInCategory())
        .orElseThrow(() -> new IllegalArgumentException("Kategorie existiert nicht!"))
```

File: backend/src/main/java/com/example/demo/soldItems/SoldItemService.java [hunk 2]
```diff
@@ -72,31 +76,46 @@ private SoldItem makeSoldItem(HelloCashItem item, HelloCashInvoice invoice) {
...
+        dateRangeToGet.add(dateStart.toString());
         while (!dateStart.equals(dateStop)) {
-            datesToGet.add(dateStart.plusDays(1L).toString());
+            dateRangeToGet.add(dateStart.plusDays(1L).toString());
             dateStart = (dateStart.plusDays(1L));
         }
 
-        for (int i = 0; i < datesToGet.size(); i++) {
-            itemsInRange.add(soldItemRepository.findAllByInvoiceTimestampContains(datesToGet.get(i)));
+        for (int i = 0; i < dateRangeToGet.size(); i++) {
```
Comment (gossie) [inline backend/src/main/java/com/example/demo/soldItems/SoldItemService.java:101]: I am pretty sure the rest of the method can be rewritten like this:
```java
return dataInRangeToGet.stream()
        .flatMap(date -> soldItemRepository.findAllByInvoiceTimestampContains(date))
        .filter(soldItem -> itemsInCategory.contains(soldItem.getItemName());
```

File: frontend/src/Main/Categories.tsx [hunk 3]
```diff
@@ -46,7 +48,7 @@ const Categories = () => {
             .then(response => {
                 if (response.status === 409) {
                     throw new Error(t("Diese Kategorie gibt es schon!"))
-                } else if (response.status ===405) {
+                } else if (response.status === 405) {
```
Comment (gossie) [inline frontend/src/Main/Categories.tsx:51]: The status 405 is "Method not allowed". In this case you would only get it, if there is no `PostMapping` in the controller. I don't think you need to handle that.

**Review by gossie — COMMENTED** [review]
Some comments.

File: backend/src/main/java/com/example/demo/soldItems/SoldItemService.java (hunk 1 above)
Comment (stedo005) [reply backend/src/main/java/com/example/demo/soldItems/SoldItemService.java:83]: How can i do a .map without a stream?

File: backend/src/main/java/com/example/demo/soldItems/SoldItemService.java (hunk 2 above)
Comment (stedo005) [reply backend/src/main/java/com/example/demo/soldItems/SoldItemService.java:101]: I try this but it´s not working

File: frontend/src/Main/Categories.tsx (hunk 3 above)
Comment (stedo005) [reply frontend/src/Main/Categories.tsx:51]: I do this for error message in frontend, it´s not allowed to create a category without a input.

**Pull request merged**